protocol/river-layout-unstable-v1.xml

```diff
@@ -0,0 +1,194 @@
+      <arg name="master_factor" type="fixed" summary="proposed master area factor"/>
+      <arg name="x" type="int" summary="x coordinate of useable area on output"/>
+      <arg name="y" type="int" summary="y coordinate of useable area on output"/>
+      <arg name="width" type="uint" summary="width of useable area on output"/>
+      <arg name="height" type="uint" summary="height of useable area on output"/>
+      <arg name="output" type="object" interface="wl_output"/>
+      <arg name="serial" type="uint" summary="serial of request"/>
+    </event>
+
+    <event name="advertise_view">
+      <description summary="make layout client aware of view">
+        This event is raised by the compositor after a request_layout event.
+        It contains additional information about one of the set of views for
+        which a layout has been requested.
+
+        It is guaranteed that every view in the layout will be advertised
+        exactly once, in an undefined order.
+
+        A client not interested in the additional information may ignore this
+        event.
+
+        The serial is the same as that of the layout request this event belongs
+        to.
+      </description>
+      <arg name="index" type="uint" summary="index of view in list"/>
+      <arg name="tag_set" type="uint" summary="bit field, active tags of view"/>
+      <arg name="app_id" type="string" summary="view app-id"/>
+      <arg name="serial" type="uint" summary="serial of layout request"/>
+    </event>
+
+    <event name="advertise_done">
+      <description summary="all views have been advertised">
+        The compositor raises this event when it has advertised all of the set
+        of views for which a layout has been requested.
+
+        A client not interested in the additional information may ignore this
+        event.
+
+        The serial is the same as that of the layout request this event belongs
+        to.
+      </description>
+      <arg name="serial" type="uint" summary="serial of layout request"/>
+    </event>
+
+    <request name="set_view_dimensions">
+      <description summary="propose view dimensions">
+        A client may use this to propose position and dimensions of one of the
+        set of views for which a layout has been requested.
+
+        It is not necessary to propose position and dimensions for every view.
+        If the client did not make a proposal of these for a view, they will be
+        determined by the compositor.
+
+        This request may be send before the view of the index has been
+        advertised by the compositor.
+
+        Dimensions are in output local logical coordinates per XDG-output
+        protocol extensions.
+
+        The serial is used to identify the layout request this request is a
+        response to.
+      </description>
+      <arg name="serial" type="uint" summary="serial of layout request"/>
+      <arg name="index" type="uint" summary="index of view in layout"/>
```

Comment:
We don't strictly need the index argument, we could simply rely on the order. We could make it a protocol error to send too many or too few `set_view_dimensions` requests.

Reply:
With the index, logic of more advanced layout clients is simplified because they can send view dimensions in any order and don't need to keep a list.

Reply:
That sounds fine to me. We need protocol errors for setting an out of bounds index and for leaving an index unset then.

Reply:
Hmm, I'm wavering on this. It makes the server code much more complex and doesn't simplify things greatly even for "advanced" clients. Such "advanced" clients likely already need to allocate an array to hold the information from advertise_view requests, I don't see needing to allocate a second to be a huge burden. They could even probably reuse the same array with a bit of cleverness.